src/nwb_conversion_tools/utils/checks.py

```diff
@@ -2,6 +2,9 @@
 
 
 def check_regular_series(series: np.ndarray, tolerance_decimals: int = 9):
```

Comment:
Following the design principle that functions should do one thing I think that you should not return the rate. Calculating the rate is only one line (9) and can be done in-line when needed (or you can have another function if you think that's better). 

Also, maybe we can have a more specific name than "regular"? Is regular a terminology that is used in `pynwb`? Maybe `check_regular_rate`?

Reply:
The function already has implicit access to the exact piece of information needed, and it is still just doing one thing (the same thing it always did).

In particular, adding an extra return value does not break any compatibility since the original usage is not changed in any way, and the extra value is informative to parsing the result.

Reply:
> The function already has implicit access to the exact piece of information needed, and it is still just doing one thing (the same thing it always did).

It is not doing one thing but two: It checks that the series has a constant rate (which is consistent with the name) and it calculate the rate (which is not consistent with the name).  

My argument is not about backwards compatibility which is fine as you are saying. Is about sticking to design standards to improve readability.

Reply:
>It is not doing one thing but two: It checks that the series has a constant rate (which is consistent with the name) and it calculate the rate (which is not consistent with the name).

And a function that calculates the rate would need to check if the series is regular. The thing you are 'doing' is checking if the series is regular. The function could likewise return the `rate` as a non-`None` value, and the boolean logic asserted as `rate is None`. The two outputs are inherently related.

Reply:
Yes. But that way you separate the two things. I thinks is way better to definine an extra function `calculate_regular_sampling_rate` that returns the sampling rate if the series is regular and and assertion (preferred) or None if it is not. That function would be doing what it tells you it does. `check_regular_series` does an extra thing that is not evidence from the name. 

I feel that something being a necessary step for something else does not indicate the two things are the same and readability trumps line reduction.

Reply:
> I feel that something being a necessary step for something else does not indicate the two things are the same and readability and clarity of intent trumps line reduction.

@h-mayorquin Please explain how

```python
is_regular, rate = check_regular_series(series=series)
```
is not both readable and minimal.

Since we're apparently at a stalemate, maybe let @weiglszonja decide between the two since it's her PR.

Reply:
I think this is good. Here is a rehash of my claims and my issue with the current approach:
* It does not follow the "functions should do one thing principle". The reason here is that the function `check_regular_series` does two things: 1) it check if the series is regular, 2) it calculates the rate. 
* The second claim has to do with readability and that function naming should reflect intent and be transparent. Right now the function `check_regular_series`  does a second thing (calculating the rate) that is not reflected in the name of the function. This makes it harder to read as it executes a non stated operation (calculating the rate) and also assumes that the users know that the rate is to be returned as the second argument (that is, an implicit contract). I think this is bad because it forces knowledge into the users and also reduces discoverability (i.e. if we had a function `calculate_series_rate` it is obvious what it does just from the name if you are looking for that functionality).
* I am proposing that we define another function called `calculate_series_rate` as this name is clear in intent. I think this is more readable and more flexible as there are situations where we might need to do the first thing `check_regular_series` but not the second `calculate_regular_rate`.  I agree that the second function might as well use the former (for example if you want to throw an assertion error or a warning) and that's fine. Functions can support each other.

Overall, my feeling is that you oppose this because you don't want to add more lines. That's why I said that "readability and clarity of intent trumps line reduction". Besides that, I am failing to imagine other argument to not separate them. Maybe I am missing another concern of yours?

Reply:
> It does not follow the "functions should do one thing principle". The reason here is that the function check_regular_series does two things: 1) it check if the series is regular, 2) it calculates the rate.

What I'm trying to show you is these two things _are_ the same thing, it's in fact how (1) determines regularity. It calculates the first-order difference (in nomenclature, 'rate' is the first-order difference of time) of the series and evaluates if they are all the same value (*i.e.*, 'constant', therefore 'regular').

Now if we have a function `calculate_series_rate` that returns a scalar when the series is regular and `None` if not, then the existence of another separate function `check_regular_series`, which is simply equivalent to the boolean evaluation `calculate_series_rate(series=series) is not None`, becomes redundant as a function unto itself, because the boolean evaluation is sufficient anywhere that logic is needed.

You say it helps discoverability, but that's heavily contingent on documentation and otherwise is a slippery slope to having a large number of functions spread across multiple projects that all evaluate similar if not identical logic, which leads to high redundancy and low efficiency, which just makes more work for ourselves regarding maintenance. That also harkens to my point of centralizing such helper functions, as a change to even a single line amplifies the amount of work we have to do across projects, which reduces our billable productivity.

Reply:
I understand how the calculation rate is done. You are saying that the two things are the same if one is required in the calculation of the other. I disagree with that. One operation is a boolean, the other returns a number. I think that's the crux of the issue. You are lumping all of that in the box of "things required to calculate the rate" which produces two things as an output and the name you use for that box is `check_regular_series`. I am advocating for more precision. 

In that regard, I am totally fine with the second approach you propose above. It appears to me that is reasonable that a `calculate_series_constant_rate` function returns None if the function is not regular and returns the rate (as expected from the name) if it exists. It avoids the need of a a function just to check for regularity and another for the rate which you want to avoid. I think that seems like a good compromise.

For the discoverability I was thinking on something simpler (doing a search function for rate in the repo which what I do when I am looking for functionality) and what a crawler will get in the read the docs. That said, I don't think that discoverability is a particularly strong argument here.

Reply:
We'll go for the revised option, then, unless @weiglszonja has a preference for the other ways

```python
def calculate_regular_series_rate(series) -> Optional[Real]:
    diff_ts = np.diff(series).round(decimals=tolerance_decimals)
    uniq_diff_ts = np.unique(diff_ts)
    rate = diff_ts[0] if len(uniq_diff_ts) == 1 else None
    return rate
```

and regularity inferred from this output being `not None`

Reply:
@CodyCBakerPhD, @h-mayorquin I read through your comments, it was very insightful, I would go for the revised option as well.